* **Mounting the Drive**

In [ ]:
# Mounting the Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


##  **1. Importing the Required Modules**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

##  **2. Importing the Dataset**

In [ ]:
df=pd.read_csv('http://data.insideairbnb.com/uconn/msbapm/capstone/2022/fall/listings-united-states-2022-07.csv.gz', compression='gzip')

* **Printing the First 5 Rows of the Dataset**

In [ ]:
pd.set_option('display.max_columns', None)
df.head()

id                                      listing_url  \
0            35575312            https://www.airbnb.com/rooms/35575312   
1  635519741001985764  https://www.airbnb.com/rooms/635519741001985764   
2            50490291            https://www.airbnb.com/rooms/50490291   
3            40991125            https://www.airbnb.com/rooms/40991125   
4            52090522            https://www.airbnb.com/rooms/52090522   

        scrape_id last_searched last_scraped  \
0  20220717002814    2022-07-17   2022-07-22   
1  20220717002814    2022-07-17   2022-07-20   
2  20220717002814    2022-07-17   2022-07-23   
3  20220717002814    2022-07-17   2022-07-27   
4  20220717002814    2022-07-17   2022-07-23   

                                                name  \
0   WV Mountain cottage minutes from New River Gorge   
1                  Ptarmigan Ranch North Fork Views!   
2                    The Northwoods Nook Blairsville   
3  Cozy Private Casita on Canyon Road, Walk to Plaza   
4  Guest Suite w/ Forest & Pasture Views Mins Fro...   

                                         description  \
0  Located on Green Acres campground, just .2 mil...   
1  Welcome to Ptarmigan Ranch! This beautiful thr...   
2  Unwind at the Northwoods Nook in the mountains...   
3  Casita de Artista is like new and a recently r...   
4  Forget your worries in this peaceful and seren...   

                               neighborhood_overview  \
0  Rural community. Quaint nearby towns and growi...   
1                                                NaN   
2                Quiet gravel road plenty of privacy   
3  Canyon Road is a historic and lovely neighborh...   
4                                                NaN   

                                         picture_url    host_id  \
0  https://a0.muscache.com/pictures/f0ee66c4-9091...   60091041   
1  https://a0.muscache.com/pictures/prohost-api/H...   80717299   
2  https://a0.muscache.com/pictures/miso/Hosting-...   22954841   
3  https://a0.muscache.com/pictures/4d4e8853-9e4f...  212710788   
4  https://a0.muscache.com/pictures/miso/Hosting-...  119778179   

                                      host_url     host_name  host_since  \
0   https://www.airbnb.com/users/show/60091041        Jeanne  2016-02-23   
1   https://www.airbnb.com/users/show/80717299  Cody Lodging  2016-06-29   
2   https://www.airbnb.com/users/show/22954841          Paul  2014-10-25   
3  https://www.airbnb.com/users/show/212710788       Jessica  2018-08-30   
4  https://www.airbnb.com/users/show/119778179       Jessica  2017-03-08   

                                   host_location  \
0  Murrells Inlet, South Carolina, United States   
1                   Cody, Wyoming, United States   
2            Blairsville, Georgia, United States   
3            Santa Fe, New Mexico, United States   
4  Hendersonville, North Carolina, United States   

                                          host_about  host_response_time  \
0             Easy going. Like to laugh and travel.       within an hour   
1  Are you looking for a true Western vacation? T...      within an hour   
2  After 30 years as a corporate chef it was time...      within an hour   
3  I’m a graduate student in the United States. I...  within a few hours   
4                                                NaN      within an hour   

  host_response_rate host_acceptance_rate host_is_superhost  \
0               100%                  90%                 t   
1                94%                  97%                 t   
2               100%                 100%                 t   
3               100%                  94%                 t   
4               100%                 100%                 f   

                                  host_thumbnail_url  \
0  https://a0.muscache.com/im/pictures/user/d1b9a...   
1  https://a0.muscache.com/im/pictures/user/b58c3...   
2  https://a0.muscache.com/im/pictures/user/cf592...   
3  https://a0.muscache.com/im/pictures/us

* **Summary of the Dataset**

In [ ]:
# Summary
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1252866 entries, 0 to 1252865
Data columns (total 80 columns):
 #   Column                                        Non-Null Count    Dtype  
---  ------                                        --------------    -----  
 0   id                                            1252866 non-null  int64  
 1   listing_url                                   1252866 non-null  object 
 2   scrape_id                                     1252866 non-null  int64  
 3   last_searched                                 1122641 non-null  object 
 4   last_scraped                                  1252866 non-null  object 
 5   name                                          1252825 non-null  object 
 6   description                                   1237927 non-null  object 
 7   neighborhood_overview                         762171 non-null   object 
 8   picture_url                                   1252863 non-null  object 
 9   host_id                            

##  **3. Data Preprocessing**

### **3.1 Exploring Missing Values**

In [ ]:
print("Number of Missing Values (Columnwise) :\n")
pd. set_option('display.max_rows', 500)
df.isna().sum()

Number of Missing Values (Columnwise) :



id                                                    0
listing_url                                           0
scrape_id                                             0
last_searched                                    130225
last_scraped                                          0
name                                                 41
description                                       14939
neighborhood_overview                            490695
picture_url                                           3
host_id                                               0
host_url                                              0
host_name                                          7750
host_since                                         7749
host_location                                     10017
host_about                                       524237
host_response_time                               150614
host_response_rate                               150614
host_acceptance_rate                            

In [ ]:
print("Total Number of Missing Values :")
df.isna().sum().sum()

Total Number of Missing Values :


12398583

### **3.2 Dropping Columns with Missing Values for every Listing**

In [ ]:
# Creating a List of Columns with Missing Values Corresponding to Every Listing
cols=list()
col_names=list(df.columns)
for i in col_names:
  if df[i].isna().sum()==df.shape[0]:
    cols.append(i)
cols

['bathrooms',
 'calendar_updated',
 'requires_license',
 'region_parent_parent_id',
 'region_parent_parent_name']

In [ ]:
df=df.drop(cols,axis=1)

### **3.3 Dropping Columns Irrelevant to the Analysis / Columns with Redundant Information**

In [ ]:
df.columns

Index(['id', 'listing_url', 'scrape_id', 'last_searched', 'last_scraped',
       'name', 'description', 'neighborhood_overview', 'picture_url',
       'host_id', 'host_url', 'host_name', 'host_since', 'host_location',
       'host_about', 'host_response_time', 'host_response_rate',
       'host_acceptance_rate', 'host_is_superhost', 'host_thumbnail_url',
       'host_picture_url', 'host_neighbourhood', 'host_listings_count',
       'host_total_listings_count', 'host_verifications',
       'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
       'latitude', 'longitude', 'property_type', 'room_type', 'accommodates',
       'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
       'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm', 'has_availability', 'availability_30',
       'availability_60', 'availabi

In [ ]:
df=df.drop(['host_thumbnail_url','host_picture_url','host_neighbourhood','neighbourhood','calendar_last_scraped','host_response_time', 'host_response_rate',
       'host_acceptance_rate','first_review', 'last_review', 'review_scores_accuracy',
       'review_scores_cleanliness', 'review_scores_checkin',
       'review_scores_communication', 'review_scores_location',
       'review_scores_value','calendar_last_scraped','bathrooms_text', 'amenities','host_listings_count',
       'host_total_listings_count','host_verifications',
       'host_has_profile_pic','region_parent_id','host_is_superhost','region_id','listing_url', 'scrape_id', 'last_searched', 'last_scraped',
       'name', 'description', 'neighborhood_overview', 'picture_url','host_url','host_about','has_availability', 'availability_30',
       'availability_60', 'availability_90', 'availability_365','minimum_minimum_nights',
       'maximum_minimum_nights', 'minimum_maximum_nights',
       'maximum_maximum_nights', 'minimum_nights_avg_ntm',
       'maximum_nights_avg_ntm'],axis=1)

### **3.4 Imputing Missing Values for 'Beds' & 'Bedrooms' columns with respective Median Values**

In [ ]:
df['beds']=df['beds'].fillna(df['beds'].median())

In [ ]:
df['bedrooms']=df['bedrooms'].fillna(df['bedrooms'].median())

## **4. Feature Engineering**

### **4.1 Adding variable to check if a listing has a license or no (If Licensed then 1 , else 0)**

In [ ]:
df['with_license']=np.where(df['license'].isna(),0,1)

In [ ]:
df=df.drop(['license'],axis=1)

### **4.2 Adding variable to check if a listing is a STR or a LTR**

In [ ]:
df['STR_vs_LTR']=np.where(df['minimum_nights']<30,'STR','LTR')

In [ ]:
df['STR_vs_LTR'].value_counts()

STR    1133932
LTR     118934
Name: STR_vs_LTR, dtype: int64

### **4.3 Adding New Columns 'estimated_booking_ltm', 'estimated_nights_ltm' & 'estimated_revenue_ltm' which represent the number of bookings, number of nights booked & revenue generated respectively for the last 12 months '**

In [ ]:
review_rate=0.5
maximum_occupancy=255
average_nights_stay=3
df['price']=df['price'].str.lstrip('$').str.replace(',','')
df['price']=pd.to_numeric(df['price'])
df['estimated_booking_ltm']=round(df['number_of_reviews_ltm']/review_rate).astype(int)
df['estimated_nights_ltm'] = round(np.minimum(df['estimated_booking_ltm'] * (np.maximum(df['minimum_nights'], average_nights_stay)),maximum_occupancy))
df['estimated_revenue_ltm'] = round(df['estimated_nights_ltm'] * df['price'])

### **4.4 Adding a New Column 'Frequency_of_renting' which shows how frequently a listing is being rented out**

In [ ]:
df.loc[df['estimated_nights_ltm']<=30, 'Frequency_of_renting'] = 'occasionally rented'
df.loc[((df['estimated_nights_ltm']>30) & (df['estimated_nights_ltm']<=90)) , 'Frequency_of_renting'] = 'semi_regularly rented'
df.loc[df['estimated_nights_ltm']>90, 'Frequency_of_renting'] = 'rented full time'

### **4.5 Adding a New Column 'Commercial Nature' with 3 Categories for the listings : 'Home Sharing', 'Semi-Commercial' & 'Commercial'**

**1. Home Sharing**

In [ ]:
df.loc[(df['room_type']=='Shared room') & (df['calculated_host_listings_count_shared_rooms']==1),'Commercial_Nature']='Home Sharing'
df.loc[(df['room_type']=='Shared room') & (df['calculated_host_listings_count_shared_rooms']>1) & (df['beds']<=3),'Commercial_Nature']='Home Sharing'
df.loc[(df['room_type']=='Private room') & (df['calculated_host_listings_count_private_rooms']<=3),'Commercial_Nature']='Home Sharing'
df.loc[(df['room_type']=='Entire home/apt') & (df['calculated_host_listings_count_entire_homes']==1) & (df['Frequency_of_renting']=='occasionally rented'),'Commercial_Nature']= 'Home Sharing'

**2. Semi-commercial**

In [ ]:
df.loc[(df['room_type']=='Shared room') & (df['calculated_host_listings_count_shared_rooms']>1) & (df['beds']>3),'Commercial_Nature']='Semi-Commercial'
df.loc[(df['room_type']=='Private room') & (df['calculated_host_listings_count_private_rooms']>3) & (df['calculated_host_listings_count_private_rooms']<=6),'Commercial_Nature']='Semi-Commercial'
df.loc[(df['room_type']=='Entire home/apt') & (df['calculated_host_listings_count_entire_homes']==1) & (df['Frequency_of_renting']=='semi_regularly rented'),'Commercial_Nature']= 'Semi-Commercial'

**3. Commercial**

In [ ]:
df.loc[(df['room_type']=='Hotel room'),'Commercial_Nature'] ='Commercial'
df.loc[(df['room_type']=='Entire home/apt') & (df['calculated_host_listings_count_entire_homes']>1),'Commercial_Nature'] ='Commercial'
df.loc[(df['room_type']=='Entire home/apt') & (df['calculated_host_listings_count_entire_homes']==1) & (df['Frequency_of_renting']=='rented full time'),'Commercial_Nature']= 'Commercial'
df.loc[(df['room_type']=='Private room') & (df['calculated_host_listings_count_private_rooms']>6),'Commercial_Nature']='Commercial'

* **Number of Listings Based on Commercial Nature**

In [ ]:
df['Commercial_Nature'].value_counts()

Commercial         854110
Home Sharing       281030
Semi-Commercial    117726
Name: Commercial_Nature, dtype: int64

## **5. Extracting Zipcodes from Latitude & Longitude Values**

In [ ]:
# Creating a New Dataframe 'df_zip'
df_zip=df[['id','latitude','longitude']]

* **Installing Required Packages**

In [ ]:
!pip install pandas fiona shapely pyproj rtree

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 16.7 MB 30.5 MB/s 
     |████████████████████████████████| 6.3 MB 68.1 MB/s 
     |████████████████████████████████| 1.0 MB 54.2 MB/s 


In [ ]:
!pip install geopandas
import geopandas as gpd

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 1.0 MB 29.8 MB/s 


In [ ]:
df_zip = gpd.GeoDataFrame(df_zip, geometry=gpd.points_from_xy(df_zip['longitude'], df_zip['latitude']),crs='EPSG:4326')

In [ ]:
zipcode_file='/content/drive/MyDrive/Advanced Business Analytics and Project Management/tl_2017_us_zcta510.zip'

In [ ]:
zipcode_gdf = gpd.read_file(f"zip://{zipcode_file}")

In [ ]:
zipcode_gdf.head()

ZCTA5CE10 GEOID10 CLASSFP10 MTFCC10 FUNCSTAT10    ALAND10  AWATER10  \
0     43451   43451        B5   G6350          S   63411475    157689   
1     43452   43452        B5   G6350          S  121783674  13437380   
2     43456   43456        B5   G6350          S    9389362    999166   
3     43457   43457        B5   G6350          S   48035540         0   
4     43458   43458        B5   G6350          S    2573816     39915   

    INTPTLAT10    INTPTLON10  \
0  +41.3183010  -083.6174935   
1  +41.5157923  -082.9809454   
2  +41.6468445  -082.8226641   
3  +41.2673266  -083.4274645   
4  +41.5304461  -083.2133648   

                                            geometry  
0  POLYGON ((-83.70873 41.32733, -83.70815 41.327...  
1  POLYGON ((-83.08698 41.53780, -83.08256 41.537...  
2  MULTIPOLYGON (((-82.83558 41.71082, -82.83515 ...  
3  POLYGON ((-83.49650 41.25371, -83.48382 41.253...  
4  POLYGON ((-83.22229 41.53102, -83.22228 41.532...

In [ ]:
zipcode_gdf.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 33144 entries, 0 to 33143
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   ZCTA5CE10   33144 non-null  object  
 1   GEOID10     33144 non-null  object  
 2   CLASSFP10   33144 non-null  object  
 3   MTFCC10     33144 non-null  object  
 4   FUNCSTAT10  33144 non-null  object  
 5   ALAND10     33144 non-null  int64   
 6   AWATER10    33144 non-null  int64   
 7   INTPTLAT10  33144 non-null  object  
 8   INTPTLON10  33144 non-null  object  
 9   geometry    33144 non-null  geometry
dtypes: geometry(1), int64(2), object(7)
memory usage: 2.5+ MB


In [ ]:
df_zip.to_crs(epsg=4269, inplace=True)
df_zip.crs

<Geographic 2D CRS: EPSG:4269>
Name: NAD83
Axis Info [ellipsoidal]:
- Lat[north]: Geodetic latitude (degree)
- Lon[east]: Geodetic longitude (degree)
Area of Use:
- name: North America - onshore and offshore: Canada - Alberta; British Columbia; Manitoba; New Brunswick; Newfoundland and Labrador; Northwest Territories; Nova Scotia; Nunavut; Ontario; Prince Edward Island; Quebec; Saskatchewan; Yukon. Puerto Rico. United States (USA) - Alabama; Alaska; Arizona; Arkansas; California; Colorado; Connecticut; Delaware; Florida; Georgia; Hawaii; Idaho; Illinois; Indiana; Iowa; Kansas; Kentucky; Louisiana; Maine; Maryland; Massachusetts; Michigan; Minnesota; Mississippi; Missouri; Montana; Nebraska; Nevada; New Hampshire; New Jersey; New Mexico; New York; North Carolina; North Dakota; Ohio; Oklahoma; Oregon; Pennsylvania; Rhode Island; South Carolina; South Dakota; Tennessee; Texas; Utah; Vermont; Virginia; Washington; West Virginia; Wisconsin; Wyoming. US Virgin Islands. British Virgin Islands

In [ ]:
zipcode_gdf.shape

(33144, 10)

In [ ]:
df_with_zips = gpd.sjoin(df_zip, zipcode_gdf, how="inner")

In [ ]:
df_with_zips.shape

(1243827, 14)

In [ ]:
df_with_zips.head()

id  latitude  longitude                    geometry  \
0                35575312  37.97405  -81.15124  POINT (-81.15124 37.97405)   
14795  621698387871328996  37.98937  -81.13412  POINT (-81.13412 37.98937)   
18490            53134724  37.98915  -81.13404  POINT (-81.13404 37.98915)   
19075  553521440303759961  37.98871  -81.13633  POINT (-81.13633 37.98871)   
33281            50914919  37.97812  -81.14890  POINT (-81.14890 37.97812)   

       index_right ZCTA5CE10 GEOID10 CLASSFP10 MTFCC10 FUNCSTAT10   ALAND10  \
0             7248     25901   25901        B5   G6350          S  85074437   
14795         7248     25901   25901        B5   G6350          S  85074437   
18490         7248     25901   25901        B5   G6350          S  85074437   
19075         7248     25901   25901        B5   G6350          S  85074437   
33281         7248     25901   25901        B5   G6350          S  85074437   

       AWATER10   INTPTLAT10    INTPTLON10  
0       1005413  +37.9507017  -081.1170495  
14795   1005413  +37.9507017  -081.1170495  
18490   1005413  +37.9507017  -081.1170495  
19075   1005413  +37.9507017  -081.1170495  
33281   1005413  +37.9507017  -081.1170495

In [ ]:
df_with_zips=df_with_zips[['id','ZCTA5CE10']]
df_with_zips.columns=['id','ZipCode']

In [ ]:
df_with_zips.head()

id ZipCode
0                35575312   25901
14795  621698387871328996   25901
18490            53134724   25901
19075  553521440303759961   25901
33281            50914919   25901

In [ ]:
df_with_zips.shape

(1243827, 2)

### **5.1 Adding Zip Codes to the DataFrame**

In [ ]:
df=pd.merge(df,df_with_zips,on='id',how='left')

In [ ]:
df['id'].value_counts()

7729074               2
663068861954821831    2
35575312              1
53584071              1
47581518              1
                     ..
44463986              1
44268442              1
17640256              1
582639942520681943    1
50170556              1
Name: id, Length: 1252866, dtype: int64

**After merging, we can see that there are duplicate records corresponding to certain ids.**

In [ ]:
# Dropping Duplicate Records
df=df.drop_duplicates(subset=['id'], keep='last')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1252866 entries, 0 to 1252867
Data columns (total 36 columns):
 #   Column                                        Non-Null Count    Dtype  
---  ------                                        --------------    -----  
 0   id                                            1252866 non-null  int64  
 1   host_id                                       1252866 non-null  int64  
 2   host_name                                     1245116 non-null  object 
 3   host_since                                    1245117 non-null  object 
 4   host_location                                 1242849 non-null  object 
 5   host_identity_verified                        1245117 non-null  object 
 6   latitude                                      1252866 non-null  float64
 7   longitude                                     1252866 non-null  float64
 8   property_type                                 1252866 non-null  object 
 9   room_type                          

In [ ]:
# Number of Missing Values for Zip Codes
df['ZipCode'].isna().sum()

9041

### **5.2 Adding a New Column 'City' to the DataFrame**

In [ ]:
df['ZipCode'] = df['ZipCode'].fillna(0).astype(int)

In [ ]:
# Reading the CSV file with 'City' & 'Zip_Code' data 
df1 = pd.read_csv('/content/drive/Shareddrives/Advanced Business Analytics and Project Management/city.csv')

In [ ]:
df1.head()

Unnamed: 0  Zip_Code          City
0           0     31018     Davisboro
1           1     31072         Pitts
2           2     31213         Macon
3           3     31302  Bloomingdale
4           4     31548     Kingsland

In [ ]:
df1 = df1.iloc[: , 1:]

In [ ]:
# Renaming the Columns
df1.columns=['ZipCode', 'City']

In [ ]:
df1['ZipCode'] = df1['ZipCode'].fillna(0).astype(int)

In [ ]:
df_final = df.set_index('ZipCode').join(df1.set_index('ZipCode'))

In [ ]:
df_final.reset_index(inplace=True)

In [ ]:
df_final['ZipCode'][df_final['City'].isna()==1].value_counts()

0        9041
98205       9
83601       2
20307       1
54934       1
84144       1
Name: ZipCode, dtype: int64

**Apart from the rows with Null Values for Zip_Codes (9041 Rows), there are certain other rows for which 'City' Data is not available despite having the Zip Codes.**

In [ ]:
# Adding 'City' data for listings where 'City' data is not available despite having Zip Codes
df_final['City'][df_final['ZipCode']==98205]='Everett'
df_final['City'][df_final['ZipCode']==20307]='Washington'
df_final['City'][df_final['ZipCode']==54934]='Eureka'
df_final['City'][df_final['ZipCode']==83601]='Atlanta'
df_final['City'][df_final['ZipCode']==84144]='Salt Lake City'

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from 

In [ ]:
# Renaming the 'Zip_Code' column
df_final = df_final.rename(columns={'ZipCode': 'Zip Code'})

## **6. Renaming Geographical Hierarchies**

In [ ]:
df_final = df_final.rename(columns={'region_parent_name': 'Region/State', 'region_name': 'Province/County'})

In [ ]:
df_final.columns

Index(['Zip Code', 'id', 'host_id', 'host_name', 'host_since', 'host_location',
       'host_identity_verified', 'latitude', 'longitude', 'property_type',
       'room_type', 'accommodates', 'bedrooms', 'beds', 'price',
       'minimum_nights', 'maximum_nights', 'number_of_reviews',
       'number_of_reviews_ltm', 'number_of_reviews_l30d',
       'review_scores_rating', 'instant_bookable',
       'calculated_host_listings_count',
       'calculated_host_listings_count_entire_homes',
       'calculated_host_listings_count_private_rooms',
       'calculated_host_listings_count_shared_rooms', 'Province/County',
       'Region/State', 'reviews_per_month', 'with_license', 'STR_vs_LTR',
       'estimated_booking_ltm', 'estimated_nights_ltm',
       'estimated_revenue_ltm', 'Frequency_of_renting', 'Commercial_Nature',
       'City'],
      dtype='object')

## **7. Adding 'Licensed_5' & 'Licensed_10' variables to show which listings are licensed (based on 5% & 10% Thresholds for Licensing Requirements)**

In [ ]:
df_c_1=pd.DataFrame(df_final.groupby(['Region/State','City']).with_license.sum())

In [ ]:
df_c_2=pd.DataFrame(df_final.groupby(['Region/State','City']).id.count())

In [ ]:
df_c_3=pd.DataFrame(df_c_1.values/df_c_2.values)

In [ ]:
df_c_3.index=df_c_2.index
df_c_3.columns=['Percentage_Licensed']

In [ ]:
df_c_3[df_c_3['Percentage_Licensed']>0.05].shape

(512, 1)

In [ ]:
df_c_4=df_c_3.copy()
df_c_4['requires_license_5']=np.where(df_c_4['Percentage_Licensed']>=0.05,1,0)
df_c_4['requires_license_10']=np.where(df_c_4['Percentage_Licensed']>=0.1,1,0)

In [ ]:
df_c_4.reset_index(inplace=True)

In [ ]:
df_c_4.head()

Region/State         City  Percentage_Licensed  requires_license_5  \
0      Alabama    Abbeville                  0.0                   0   
1      Alabama   Adamsville                  0.0                   0   
2      Alabama        Akron                  0.0                   0   
3      Alabama    Alabaster                  0.0                   0   
4      Alabama  Albertville                  0.0                   0   

   requires_license_10  
0                    0  
1                    0  
2                    0  
3                    0  
4                    0

In [ ]:
df_fin=df_final.set_index(['Region/State','City']).join(df_c_4.set_index(['Region/State','City']))

In [ ]:
df_fin.reset_index(inplace=True)

In [ ]:
df_fin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1252866 entries, 0 to 1252865
Data columns (total 40 columns):
 #   Column                                        Non-Null Count    Dtype  
---  ------                                        --------------    -----  
 0   Region/State                                  1252866 non-null  object 
 1   City                                          1243825 non-null  object 
 2   Zip Code                                      1252866 non-null  int64  
 3   id                                            1252866 non-null  int64  
 4   host_id                                       1252866 non-null  int64  
 5   host_name                                     1245116 non-null  object 
 6   host_since                                    1245117 non-null  object 
 7   host_location                                 1242849 non-null  object 
 8   host_identity_verified                        1245117 non-null  object 
 9   latitude                           

In [ ]:
# For 5% Threshold
df_fin.loc[(df_fin['with_license']==1),'licensed_5']="Licensed"
df_fin.loc[(df_fin['with_license']==0) & (df_fin['requires_license_5']==0),'licensed_5']="Exempt"
df_fin.loc[(df_fin['with_license']==0) & (df_fin['requires_license_5']==1),'licensed_5']="Unlicensed"

In [ ]:
# For 10% Threshold
df_fin.loc[(df_fin['with_license']==1),'licensed_10']="Licensed"
df_fin.loc[(df_fin['with_license']==0) & (df_fin['requires_license_10']==0),'licensed_10']="Exempt"
df_fin.loc[(df_fin['with_license']==0) & (df_fin['requires_license_10']==1),'licensed_10']="Unlicensed"

In [ ]:
df_fin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1252866 entries, 0 to 1252865
Data columns (total 42 columns):
 #   Column                                        Non-Null Count    Dtype  
---  ------                                        --------------    -----  
 0   Region/State                                  1252866 non-null  object 
 1   City                                          1243825 non-null  object 
 2   Zip Code                                      1252866 non-null  int64  
 3   id                                            1252866 non-null  int64  
 4   host_id                                       1252866 non-null  int64  
 5   host_name                                     1245116 non-null  object 
 6   host_since                                    1245117 non-null  object 
 7   host_location                                 1242849 non-null  object 
 8   host_identity_verified                        1245117 non-null  object 
 9   latitude                           

### **7.1 Adding a New Column 'Country' to the Dataset**

In [ ]:
df_fin['Country']="USA"

In [ ]:
df_fin.head()

Region/State       City  Zip Code                  id    host_id host_name  \
0      Alabama  Abbeville     36310  639981864257087728   73476304       J J   
1      Alabama  Abbeville     36310  553844240661724217   53828134     Cliff   
2      Alabama  Abbeville     36310            28434319   60733647   Harriet   
3      Alabama  Abbeville     36310            53280278   65348393      Jill   
4      Alabama  Abbeville     36310  660693730581970133  466902150      Troy   

   host_since                      host_location host_identity_verified  \
0  2016-05-23                                 US                      t   
1  2016-01-08    Alapaha, Georgia, United States                      f   
2  2016-02-28       Mims, Florida, United States                      t   
3  2016-04-01  Abbeville, Alabama, United States                      f   
4  2022-06-30                                 US                      f   

    latitude  longitude     property_type        room_type  accommodates  \
0  31.765610 -85.134766  Entire townhouse  Entire home/apt             4   
1  31.662300 -85.117170       Entire home  Entire home/apt             4   
2  31.697512 -85.146591    Entire cottage  Entire home/apt             6   
3  31.557090 -85.185780    Entire cottage  Entire home/apt             2   
4  31.707760 -85.159200       Entire home  Entire home/apt             5   

   bedrooms  beds  price  minimum_nights  maximum_nights  number_of_reviews  \
0       2.0   3.0  145.0               1            1125                  3   
1       2.0   2.0   90.0               2              10                  5   
2       2.0   3.0  102.0               1            1125                115   
3       1.0   2.0   70.0               3              30                  0   
4       2.0   3.0   74.0               3             365                  0   

   number_of_reviews_ltm  number_of_reviews_l30d  review_scores_rating  \
0                      3                       3                  5.00   
1                      5                       0                  5.00   
2                     44                       3                  4.97   
3                      0                       0                   NaN   
4                      0                       0                   NaN   

  instant_bookable  calculated_host_listings_count  \
0                f                               1   
1                t                               1   
2                f                               1   
3                t                               1   
4                t                               1   

   calculated_host_listings_count_entire_homes  \
0                                            1   
1                                            1   
2                                            1   
3                                            1   
4                                            1   

   calculated_host_listings_count_private_rooms  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

   calculated_host_listings_count_shared_rooms Province/County  \
0                                            0           Henry   
1                                            0           Henry   
2                                            0           Henry   
3                                            0           Henry   
4                                            0           Henry   

   reviews_per_month  with_license STR_vs_LTR  estimated_booking_ltm  \
0               3.00             0        STR                      6   
1               1.15             0        STR                     10   
2               3.00             0        STR                     88   
3                NaN             0        STR                      0   
4  

In [ ]:
df_fin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1252866 entries, 0 to 1252865
Data columns (total 43 columns):
 #   Column                                        Non-Null Count    Dtype  
---  ------                                        --------------    -----  
 0   Region/State                                  1252866 non-null  object 
 1   City                                          1243825 non-null  object 
 2   Zip Code                                      1252866 non-null  int64  
 3   id                                            1252866 non-null  int64  
 4   host_id                                       1252866 non-null  int64  
 5   host_name                                     1245116 non-null  object 
 6   host_since                                    1245117 non-null  object 
 7   host_location                                 1242849 non-null  object 
 8   host_identity_verified                        1245117 non-null  object 
 9   latitude                           

* **Exporting the Cleaned .csv File**

In [ ]:
df_fin.to_csv('/content/drive/MyDrive/Advanced Business Analytics and Project Management/Final Files/USA.csv')